plugins/streaming/impls/mqtt/src/main/java/org/pentaho/di/trans/step/mqtt/MQTTProducer.java

```diff
@@ -191,15 +193,25 @@ private String getTopic( Object[] row ) {
     super.stopRunning( stepMetaInterface, stepDataInterface );
   }
 
+  @Override public void dispose( StepMetaInterface smi, StepDataInterface sdi ) {
+    super.dispose( smi, sdi );
+    stopMqttClient();
+  }
+
   private void stopMqttClient() {
     try {
       // Check if connected so subsequent calls does not produce an already stopped exception
-      if ( null != client.get() && client.get().isConnected() ) {
+      if ( !connectionError.get()
```

Comment:
Is there a case when connectionError and isConnected are both true?  not sure I understand what connectionError is buying us.

Reply:
connectionError is just preventing a re-attempt at making the connection.  `client` is a supplier, so if the supplier fails to initialize the first time we'll just try again when we're in the process of closing.
Not strictly necessary, but prevents some duplicate error logging.